File: Sources/ida/contest.swift
import Cocoa

struct Country: Equatable, Hashable {
    let forum: String
    let iso: String
    let names: [String]

    init(forum: String, iso: String, names: [String]) {
        self.forum = forum
        self.iso = iso
        self.names = names
    }

    func primaryName() -> String {
        self.names.first ?? ""
    }

    func containsName(line: String) -> Bool {
        self.names.contains { name in
            return line.lowercased().contains(name.lowercased())
        }
    }
}

struct Entry: Hashable {
    let country: Country
    let artist: String
    let song: String

    init(country: Country, artist: String, song: String) {
        self.country = country
        self.artist = artist
        self.song = song
    }

    func flag() -> String {
        "World/\(self.country.iso).png"
    }
}

struct Vote: Hashable {
    let entry: Entry
    let points: Int
}

struct Contest {
    let countries: [Country]
    let entries: [Entry]
    let voters: [Country]

    init(countries: [Country], entries: [Entry], voters: [Country]) {
        self.countries = countries
        self.entries = entries
        self.voters = voters
    }

    func findVoterByCountryName(countryName: String) -> Country? {
        self.voters.first { voter in
            voter.containsName(line: countryName)
        }
    }

    func findEntryByCountryName(countryName: String) -> Entry? {
        self.entries.first { entry in
            entry.country.containsName(line: countryName)
        }
    }

    func findEntryByArtist(line: String) -> Entry? {
        self.entries.first { entry in
            line.lowercased().contains(entry.artist.lowercased())
        }
    }

    func findEntryBySong(line: String) -> Entry? {
        self.entries.first { entry in
            line.lowercased().contains(entry.song.lowercased())
        }
    }

    func findEntryByCountryForum(line: String) -> Entry? {
        self.entries.first { entry in
            line.lowercased().contains(entry.country.forum.lowercased())
        }
    }

    func copyVotesToClipboard(voter: Country?, votes: [Vote]) {
        var votesList = Array(repeating: "", count: self.entries.count)
        if voter != nil {
            if let index = self.entries.map({$0.country}).firstIndex(of: voter) {
                votesList[index] = "X"
            }
        }

        for vote in votes {
            if let index = self.entries.firstIndex(of: vote.entry) {
                votesList[index] = String(vote.points)
            }
        }

        let toCopy = votesList.joined(separator: "\n")
        let pasteBoard = NSPasteboard.general
        pasteBoard.clearContents()
        pasteBoard.setString(toCopy, forType: .string)
    }
}

File: src/contest.ts
import { loadJSON, loadTextLines } from "./utilities";

export class Country {
    forum: string;
    names: string[];
    flag?: string;

    constructor(forum: string, names: string[], flag?: string) {
        this.forum = forum;
        this.names = names;
        this.flag = flag;
    }

    static fromFile(path: string): Country[] {
        interface CountriesFile {
            countries: Country[]
        }

        const data = loadJSON(path) as CountriesFile;
        return data.countries.map(c => new Country(c.forum, c.names, c.flag));
    }

    primaryName(): string | undefined {
        if (this.names.length > 0) {
            return this.names[0];
        }
        return undefined;
    }

    containedIn(text: string): boolean {
        return this.names.some(name => text.toLowerCase().includes(name.toLowerCase()));
    }
}


export class Entry {
    country: Country;
    artist: string;
    song: string;

    constructor(country: Country, artist: string, song: string) {
        this.country = country;
        this.artist = artist;
        this.song = song;
    }

    static fromFile(path: string, countries: Country[]): Entry[] {
        const countriesMap: Record<string, Country> = {};
        countries.forEach(country => countriesMap[country.forum] = country);
        return loadTextLines(path)
            .filter(line => line) // empty strings '' evaluate to false and is filtered out
            .map(line => {
                const forum = line.match(/:([A-z]*):/)![0].replace(/:/g, '');
                const country: Country = countriesMap[forum];
                const [artist, song] = line.match(/\[B\](.*)\[\/B\]/)![0]
                    .replace('[B]', '').replace('[/B]', '')
                    .split(' - ')
                return new Entry(country, artist, song);
            });
    }

    flag(): string {
        if (this.country.flag === undefined) { return ''; }
        else { return `World/${this.country.flag}.png`; }
    }
}

export class Vote {
    constructor(
        public entry: Entry,
        public points: number
    ) { }
}

export class Contest {
    entries: Entry[];
    countries: Country[];
    voters: Country[];

    constructor(entries: Entry[], countries: Country[], voters: Country[]) {
        this.entries = entries;
        this.countries = countries;
        this.voters = voters;
    }

    static fromFile(path: string): Contest {
        const data = loadJSON(path) as Contest;
        const entries = data.entries.map(e => new Entry(new Country(e.country.forum, e.country.names, e.country.flag), e.artist, e.song));
        const countries = data.countries.map(c => new Country(c.forum, c.names, c.flag));
        const voters = data.voters.map(c => new Country(c.forum, c.names, c.flag));
        return new Contest(entries, countries, voters);
    }

    findVoterByCountryName(voterName: string): Country | undefined {
        return this.voters.find(voter => voter.containedIn(voterName));
    }

    findEntryByCountryName(text: string): Entry | undefined {
        return this.entries.find(entry => entry.country.containedIn(text));
    }

    findEntryByCountryForum(text: string): Entry | undefined {
        return this.entries.find(entry => text.toLowerCase().includes(entry.country.forum.toLowerCase()));
    }

    findEntryByArtist(text: string): Entry | undefined {
        return this.entries.find(entry => text.toLowerCase().includes(entry.artist.toLowerCase()));
    }

    findEntryBySong(text: string): Entry | undefined {
        return this.entries.find(entry => text.toLowerCase().includes(entry.song.toLowerCase()));
    }
}